File: src/main/java/botmanager/frostbalance/GuildWrapper.java
package botmanager.frostbalance;

import botmanager.Utilities;
import botmanager.frostbalance.data.RegimeData;
import botmanager.frostbalance.data.TerminationCondition;
import botmanager.frostbalance.grid.WorldMap;
import lombok.Getter;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.exceptions.ErrorResponseException;
import net.dv8tion.jda.api.exceptions.InsufficientPermissionException;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.List;
import java.util.*;

public class GuildWrapper {

    @Getter
    transient Frostbalance bot;

    @Getter
    String guildId;

    transient BufferedImage guildIcon;
    @Getter
    String lastKnownName;

    @Getter
    Optional<String> ownerId = Optional.empty();

    Optional<WorldMap> map = Optional.empty();

    Set<OptionFlag> optionFlags = new HashSet<>();
    List<RegimeData> regimes = new ArrayList<>();

    public GuildWrapper(Frostbalance bot, Guild guild) {
        this.bot = bot;
        this.guildId = guild.getId();
    }

    public String getName() {
        if (getJDA().getGuildById(guildId) != null) {
            lastKnownName = getJDA().getGuildById(guildId).getName();
        }
        return lastKnownName;
    }

    public Optional<Guild> getGuild() {
        return Optional.ofNullable(getJDA().getGuildById(guildId));
    }

    public JDA getJDA() {
        return bot.getJDA();
    }

    public Color getColor() {
        if (optionFlags.contains(OptionFlag.RED)) {
            return Color.RED;
        } else if (optionFlags.contains(OptionFlag.GREEN)) {
            return Color.GREEN;
        } else if (optionFlags.contains(OptionFlag.BLUE)) {
            return Color.BLUE;
        } else {
            return Color.LIGHT_GRAY;
        }
    }

    /**
     * Returns a modifiable clone of this server's records.
     */
    public List<RegimeData> readRecords() {
        return new ArrayList<>(regimes);
    }

    public boolean hasBeenForciblyRemoved(String userId) {
        return (regimes.size() > 0 &&
                regimes.get(regimes.size() - 1).getUserId().equals(userId) &&
                regimes.get(regimes.size() - 1).getTerminationCondition().equals(TerminationCondition.RESET));
    }

    public void doCoup(User user) {
        endRegime(TerminationCondition.COUP);
        startRegime(user);
    }

    private void endRegime(TerminationCondition condition) {

        if (!getGuild().isPresent()) {
            throw new IllegalStateException("Tried to end the regime of a server the bot isn't in!");
        }

        getLeaderAsMember().ifPresent(
                leader -> getGuild().get().removeRoleFromMember(leader, getLeaderRole()).queue()
        );

        if (ownerId.isPresent()) {
            System.out.println("Ending active regime of " + ownerId);
            try {
                int lastRegimeIndex = regimes.size() - 1;
                regimes.get(lastRegimeIndex).end(condition);
            } catch (IndexOutOfBoundsException e) {
                System.err.println("Index out of bounds when trying to adjust the last regime! The history data may be lost.");
                System.err.println("Creating a fragmented history.");
                RegimeData regime = new RegimeData(this, ownerId.get());
                regime.end(condition);
                regimes.add(regime);
            }

            ownerId = Optional.empty();
        }
    }

    private void startRegime(User user) {
        RegimeData regime = new RegimeData(this, user.getId(), Utilities.todayAsLong());
        regimes.add(regime);

        getGuild().get().addRoleToMember(user.getId(), getLeaderRole()).queue();
        ownerId = Optional.of(user.getId());
    }

    private Role getLeaderRole() {
        try {
            return getGuild().get().getRolesByName("LEADER", true).get(0);
        } catch (IndexOutOfBoundsException e) {
            System.err.println(getName() + " doesn't have a valid leader role! Attempting to create...");
            Role role = getGuild().get().createRole()
                    .setColor(getColor())
                    .setName("Leader")
                    .setPermissions(Permission.ADMINISTRATOR)
                    .setHoisted(true)
                    .complete();
            getGuild().get()
                    .modifyRolePositions()
                    .selectPosition(role)
                    .moveTo(getSystemRole().getPosition() - 1)
                    .queue();
            return role;
        }
    }

    private Role getSystemRole() {
        try {
            return getGuild().get().getRolesByName("FROSTBALANCE", true).get(0);
        } catch (IndexOutOfBoundsException indexException) {
            System.err.println(getName() + " doesn't have a valid system role! Attempting to create...");
            try {
                Role role = getGuild().get().createRole()
                        .setColor(getColor())
                        .setName("Leader")
                        .setPermissions(Permission.ADMINISTRATOR)
                        .setHoisted(true)
                        .complete();
                getGuild().get().addRoleToMember(getJDA().getSelfUser().getId(), role).queue();
                getGuild().get()
                        .modifyRolePositions()
                        .selectPosition(role)
                        .moveTo(getGuild().get().getRoles().size() - 1)
                        .queue();
                return role;
            } catch (ErrorResponseException | IllegalArgumentException | InsufficientPermissionException | IllegalStateException errorException) {
                getGuild().get().getDefaultChannel().sendMessage(getGuild().get().getOwner().getAsMention() + " I cannot function because there is no 'Frostbalance' role. This server is dead without one.");
                errorException.addSuppressed(indexException);
                throw errorException;
            }
        }
    }

    private Optional<Member> getLeaderAsMember() {
        return ownerId.flatMap(ownerId -> bot.getMemberWrapper(ownerId, guildId).getMember());
    }
}

File: src/main/java/botmanager/frostbalance/GuildWrapper.kt
package botmanager.frostbalance

import botmanager.Utilities
import botmanager.frostbalance.Frostbalance.Companion.bot
import botmanager.frostbalance.flags.GuildOptions
import botmanager.frostbalance.records.RegimeData
import botmanager.frostbalance.records.TerminationCondition
import botmanager.frostbalance.flags.OldOptionFlag
import botmanager.frostbalance.grid.Containable
import botmanager.frostbalance.grid.Container
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Role
import java.awt.Color
import java.awt.image.BufferedImage
import java.util.*

class GuildWrapper(@Transient var gameNetwork: GameNetwork, var id: String) : Containable<GameNetwork>, Container {


    constructor(gameNetwork: GameNetwork, guild: Guild) : this(gameNetwork, guild.id)

    var guildOptions = GuildOptions(this)

    val members: Collection<MemberWrapper>
        get() = bot.userWrappers.mapNotNull { user -> user.memberIfIn(this) }

    @Transient
    var guildIcon: BufferedImage? = null

    @JvmField
    var lastKnownName: String? = null

    var leaderId: String? = null

    var nation: Nation =
            Nation.values().toMutableSet().subtract(gameNetwork.nations).randomOrNull() ?:
            throw IllegalStateException("This network is at capacity and cannot support another guild!")

    @JvmField
    var regimes: MutableList<RegimeData> = ArrayList()
    val name: String?
        get() {
            if (jda.getGuildById(id) != null) {
                lastKnownName = jda.getGuildById(id)!!.name
            }
            return lastKnownName
        }

    val jdaGuild: Guild?
        get() = jda.getGuildById(id)
    val jda: JDA
        get() = bot.jda
    val color: Color
        get() = nation.color

    override fun toString() : String = if (isOnline()) name!! else name?.let { "~~$it~~" }?:"*Inaccessible Guild*"

    override fun adopt() {
        regimes.forEach { regime -> regime.setParent(this) }
        if (guildOptions == null) {
            guildOptions = GuildOptions(this)
        }
        guildOptions.setParent(this)
    }

    fun isOnline(): Boolean {
        return jdaGuild != null
    }

    fun getMember(user: UserWrapper): MemberWrapper {
        return user.memberIn(id)
    }

    /**
     * Returns a modifiable clone of this server's records.
     */
    fun readRecords(): List<RegimeData> {
        return ArrayList(regimes)
    }

    fun hasBeenForciblyRemoved(userId: String): Boolean {
        return regimes.size > 0 && regimes[regimes.size - 1].userId == userId && regimes[regimes.size - 1].terminationCondition == TerminationCondition.RESET
    }

    fun doCoup(member: Member) {
        endRegime(TerminationCondition.COUP)
        startRegime(member)
    }

    fun inaugurate(member: Member) {
        endRegime(TerminationCondition.TRANSFER)
        startRegime(member)
    }

    fun markLeaderAsDeserter() {
        endRegime(TerminationCondition.LEFT)
    }

    private fun endRegime(condition: TerminationCondition) {
        check(jdaGuild != null) { "Tried to end the regime of a server the bot isn't in!" }
        leaderAsMember?.let { jdaGuild?.removeRoleFromMember(it, leaderRole)?.queue() }
        if (leaderId != null) {
            println("Ending active regime of $leaderId")
            try {
                val lastRegimeIndex = regimes.size - 1
                regimes[lastRegimeIndex].end(condition)
            } catch (e: IndexOutOfBoundsException) {
                System.err.println("Index out of bounds when trying to adjust the last regime! The history data may be lost.")
                System.err.println("Creating a fragmented history.")
                val regime = RegimeData(this, leaderId)
                regime.end(condition)
                regimes.add(regime)
            }
            leaderId = null
        }
    }

    private fun startRegime(member: Member) {
        val regime = RegimeData(this, member.id, Utilities.todayAsLong())
        regimes.add(regime)
        jdaGuild?.addRoleToMember(member.id, leaderRole)?.queue()
        leaderId = member.id
    }

    fun loadLegacy(settings: MutableSet<OldOptionFlag>, Records: MutableList<RegimeData>, OwnerId: String, Name: String) {
        nation = settings.legacyNation ?: Nation.LIGHT
        regimes = Records
        leaderId = OwnerId
        lastKnownName = Name
    }

    fun reset() {
        endRegime(TerminationCondition.RESET)
        softReset()
    }

    private fun softReset() {
        val roles = jdaGuild!!.roles
        for (role in roles) {
            if (systemRole != role && leaderRole != role) {
                role.delete()
            }
        }
        for (ban in jdaGuild!!.retrieveBanList().complete()) {
            if (!bot.getUserWrapper(ban.user.id).memberIn(id).banned)
            jdaGuild!!.unban(ban.user)
        }
    }

    val leaderRole: Role
        get() = {
            if (jdaGuild == null) {
                throw IllegalStateException("Tried to get the leader role in a non-existent guild!")
            }
            var roles = jdaGuild!!.getRolesByName("LEADER", true)
            if (roles.isEmpty()) {
                System.err.println("$name doesn't have a valid leader role! Attempting to create...")
                val role = jdaGuild!!.createRole()
                        .setColor(color)
                        .setName("Leader")
                        .setPermissions(Permission.ADMINISTRATOR)
                        .setHoisted(true)
                        .complete()
                jdaGuild!!.modifyRolePositions()
                        .selectPosition(role)
                        .moveTo(systemRole.position - 1)
                        .queue()
                role!!
            } else {
                roles[0]
            }
        }.invoke()

    val systemRole: Role
        get() = {
            if (jdaGuild == null) {
                throw IllegalStateException("Tried to get the leader role in a non-existent guild!")
            }
            var roles = jdaGuild!!.getRolesByName("FROSTBALANCE", true)
            if (roles.isEmpty()) {
                System.err.println("$name doesn't have a valid system role! Attempting to create...")
                val role = jdaGuild!!.createRole()
                        .setColor(color)
                        .setName("Leader")
                        .setPermissions(Permission.ALL_PERMISSIONS - Permission.ADMINISTRATOR.offset - Permission.BAN_MEMBERS.offset)
                        .setHoisted(true)
                        .complete()
                jdaGuild!!.addRoleToMember(jda.selfUser.id, role).queue()
                jdaGuild!!
                        .modifyRolePositions()
                        .selectPosition(role)
                        .moveTo(jdaGuild!!.roles.size - 1)
                        .queue()
                role
            } else {
                roles[0]
            }
        }.invoke()

    val leaderAsMember: Member?
        get() = leaderId?.let { bot.getMemberWrapper(it, id).jdaMember }

    companion object {
        val Guild.wrapper: GuildWrapper
            get() = bot.getGuildWrapper(id)
    }

    override fun setParent(parent: GameNetwork) {
        gameNetwork = parent
    }

    fun moveToNetwork(selectedNetwork: GameNetwork) {
        gameNetwork.moveGuild(this, selectedNetwork)
    }

    fun commandContextFooter(): String {
        return "Command executed in $name"
    }

    fun grantedContextFooter(): String {
        return "From $name"
    }

    fun getMemberByName(name: String): MemberWrapper? {
        return members.firstOrNull { member -> member.effectiveName.equals(name, ignoreCase = true)}
    }

    fun allows(player: Player?): Boolean {
        return player?.allegiance.let {
            guildOptions.openBordersWith(it) || it == nation
        }
    }

    val notAllowed: String
        get() {
            val initial = ":passport_control: $name has closed its borders."
            val exceptions = if (guildOptions.borderTreaties!!.isNotEmpty()) {
                "You cannot gain influence unless your allegiance is to one of these nations: ${toString()}, ${
                    guildOptions.borderTreaties!!.joinToString(", ") {
                        gameNetwork.guildWithAllegiance(it).toString()
                    }
                }"
            } else {
                "You cannot gain influence unless you have allegiance here."
            }
            return "$initial $exceptions"
        }

    private val MutableSet<OldOptionFlag>.legacyNation: Nation?
        get() = when {
            contains(OldOptionFlag.BLUE) -> {
                Nation.BLUE
            }
            contains(OldOptionFlag.RED) -> {
                Nation.RED
            }
            contains(OldOptionFlag.GREEN) -> {
                Nation.GREEN
            }
            else -> {
                null
            }
        }


}